x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/QueryPlan.java

```diff
@@ -48,11 +48,15 @@ public AttributeSet outputSet() {
 
     public AttributeSet inputSet() {
         if (lazyInputSet == null) {
-            List<Attribute> attrs = new ArrayList<>();
+            int size = 0;
             for (PlanType child : children()) {
-                attrs.addAll(child.output());
+                size += child.output().size();
```

Comment:
This is only better performing if the return value of `child.output()` is internally stored and re-used. I think this is the case generally, but we have no way to ensuring this properly.

If we want to use this approach, I think we should add to the javadoc of `QueryPlan.output()` that multiple calls should return the very same object.

Alternatively, `QueryPlan.outputSet` is not generally overridden by inheritors and actually stores the output set for later retrieval, providing this guarantee from the get-go. What's more, in case of unary plans (the most common plan type) _that are not leafs_, the input set is _exactly the same_ as the child's outputset, so we could avoiding any sort of computation here, at all.